Enterprise-CMCS/macpro-mdct-mcr
# Make number validation more strict

### Description
This change ensures that all fields of type `number` will have well-formatted values.

Every numeric value goes through 5 layers of validation:
1. We use a CMS Design System input for number fields. If the `mask` property is set to a value CMS recognizes, it will be auto-formatted on blur. This happens before our `onBlur` handler can fire, and happens _only_ if the value contains a digit<sup>1</sup>.
3. We have a set of custom masks (`comma-separated`, `ratio`, and `percentage`) that execute on blur. These will auto-format the fields that CMS doesn't. This also happens _only_ if the value contains a digit.
4. We run the value through our `yup`-based schema validation.
5. When it hits the backend, we run it through another `yup` schema for validity
6. And we run it through _another_ schema for completeness / statusing

This commit changes all 3 `yup` schemas from `must contain a digit, comma, or dot` to `must contain a digit, or be a single dot`. Any number that was auto-formatted by a mask in step 1 or 2 will definitely pass this check. Any string that sneaked by the masking by containing no digits will instantly show an error to the user; the only way to proceed will be to enter a digit, which will then trigger the masking<sup>2</sup>.

A single dot is also allowed, so that a user can enter `.5` without an angry error message flashing before they type the second character. In theory we could change the masking so that it also fires here, but that change would be a lot more involved than it sounds. Frankly, I don't think it would be worth doing: the risk of users entering only a dot seems very low, and low-impact, in comparison to the risk of me making a mistake while changing the masking logic. Regardless, a single dot would be allowed today, so this change is at least not doing any harm.

<sup>1</sup> For certain masks, such as `zip`, this happens only if the value contains _several_ digits. But we don't use those masks in our application, as of today. https://design.cms.gov/components/text-field/masked-field/

<sup>2</sup> This is only true since we don't use masks such as `zip`. It is possible we could use such masks in the future, so this is a gap in today's work. Our current setup makes guaranteeing a different number of digits for different input types a bit of a challenge, so I am fine with this gap for now.

### Related ticket(s)
MDCT-2391

---
### How to test
1. Find a numeric field somewhere in MCPAR or MLR.
2. Try to enter some different kinds of strings in there.
   * If you enter any string that doesn't contain a digit, you should see errors that prevent saving.
   * If you enter any poorly-formatted number, you should see it reformatted on blur.

One edge case worth mentioning is the sanction dollar amount in MCPAR section D. This is the only numeric field in MCPAR which uses a built-in CMS mask (`currency`) rather than a custom mask.

### Important updates
n/a

---
### Author checklist
